### DISCUSSION TITLE
Parallelization Questions

### DISCUSSION THREAD

**Discussion by ProfessorManhattan** [go-task/task#700] [Q&A]
Hey, I have a few questions:

1. In a for or while loop, how should we be calling other tasks? Calling tasks with the `task taskname` starts another process and all of the `run: once` are reset. Also, how can we run all the tasks in parallel? Is it recommended to run `task taskname &`?
2. In the `cmds` block, how can we run, say, 3 different tasks in parallel?
3. If a task has 3 deps and those 3 deps all have a couple `- task: taskname` lines in their `cmds` block, will all 3 deps still run in parallel (including their secondary task calls)?
4. If 3 tasks running in parallel call the same task, is there a way of queueing the "same task" so that it only run consecutively? This is needed in cases like an `apt-get` where only one instance can run at a time.

**Comment by andreynering** [comment]
Hi @ProfessorManhattan,

There's no ideal answer at the moment. Calling `task` binary inside the Taskfile is not ideal, but will work for simple use cases.

We're a bit overdue on implementing builtin support for looping files, etc. There are some old issues about it like #82 and #140, so it may eventually happen. We're planning v4 on #694, but I can't promise this will be implemented.

**Reply by ghostsquad** [reply]
@ProfessorManhattan may I ask what your use case is of running the `task` binary instead of just referring to a task?

**Reply by ProfessorManhattan** [reply]
Hey @ghostsquad -- I have countless instances where I need to use task within a loop. I am basically using Task for all scripts since it's easy to run things in parallel and it makes software dependency installation at runtime possible. You can check out my Taskfiles here in this project which is an interesting, albeit not yet released fork of Task: https://github.com/megabyte-labs/Bodega -- the taskfiles are in Taskfile.yml and in .config/taskfiles --- I wouldn't be surprised if my taskfiles are the bulkiest single collection... would love to hear criticisms from you especially since you are a Collaborator

**Reply by ghostsquad** [reply]
@ProfessorManhattan thank you for sharing! I love seeing implementations by power users. It will really help inform me how `task` is used out in the wild, and determine what features are most needed.

**Reply by ghostsquad** [reply]
@ProfessorManhattan do you have a specific task/taskfile you can point me at that calls out the `task` binary? So I can see what you are trying to do with it there?

**Reply by ghostsquad** [reply]
> Bodega takes go-task to the next level by improving the TUI experience and including features that allow you to use the project as a CLI-generator.

🤣 oh boy... I'm doing the same thing with v4 of task.

This certainly is quite the `Taskfile.yml` you have.

**Reply by ProfessorManhattan** [reply]
@ghostsquad -- let's collaborate

**Reply by ProfessorManhattan** [reply]
Hey @ghostsquad -- I'd say the gnarliest circumstance I can think of is where I'm running task with a routine that includes committing code with git but then the Husky pre-commit git hooks trigger and also call a task. There are also maybe a dozen or two loops that include task... you can find them by searching for "task " with a space after it. Here's an example of one: https://gitlab.com/megabyte-labs/common/shared/-/blob/master/.config/taskfiles/docker/Taskfile-test.yml#L93

**Comment by ghostsquad** [comment]
I can try to answer some of your questions, as well as give you some insight into how I would like to address some of these things in v4.

> ❓ In a for or while loop, how should we be calling other tasks? Calling tasks with the `task taskname` starts another process and all of the `run: once` are reset. Also, how can we run all the tasks in parallel? Is it recommended to run `task taskname &`?

It seems you want to run some sub-task or a step an indeterminate number of times, of which you would know only at runtime?
As of task v3, there's no support for loops in this way. In v4 I am looking to introduce several features that would likely solve this problem:

1. The introduction of `matrix` functionality, similar to that in [github actions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) - I'm currently considering 2 implementations of this:  [task matrix](https://github.com/ghostsquad/task/blob/feat/v4/sdk/v4/taskfile/task.go#L128-L156) and [step matrix](https://github.com/ghostsquad/task/blob/feat/v4/sdk/v4/taskfile/step.go#L18-L42)

2. Dynamic Fields - This is not yet documented, as I'm still mulling over how I want this to behave, but basically, anywhere in the Taskfile.yml where you can provide a `list` or `map`, I'll be including an option for you to provide a `template`, which needs to return `json`, which is then interpreted as if that value was the value of the original field. Example

```yaml
tasks:
   a:
     do:
     - echo "a"
   b:
     do:
     - echo "b"
   c:
     do:
     - echo "c"
   ham:
     needs:
       tpl: '{{ list "a" "b" "c" | mustToJson }}'
```

If indeed you don't know exactly what dependencies you need to run until runtime, this may be able to solve that.

3. In v4, Each task will have a field called `serialGroups` which allows you to more liberally put all your dependent tasks inside `deps` (renamed to `needs` in v4) and still control parallelization. [more info](https://github.com/ghostsquad/task/blob/feat/v4/sdk/v4/taskfile/task.go#L115-L117) • [inspired by](https://concourse-ci.org/jobs.html#schema.job.serial_groups)

---

> ❓ In the `cmds` block, how can we run, say, 3 different tasks in parallel?

In both v3 and v4, you can define a `task` to run within `cmds` (renamed to `do` in v4), and that task can have `deps` (renamed to `needs` in v4). Those `needs` (`deps`) run in parallel.

---


>  ❓ If a task has 3 deps and those 3 deps all have a couple `- task: taskname` lines in their `cmds` block, will all 3 deps still run in parallel (including their secondary task calls)?

@andreynering may know the answer off hand, I'm still getting intimate with the code base. But according to [the documentation](https://taskfile.dev/#/usage?id=limiting-when-tasks-run)

> If a task executed by multiple cmds or multiple deps you can control when it is executed using run. run can also be set at the root of the Taskfile to change the behavior of all the tasks unless explicitly overridden.

> Supported values for run:
>
> - `always` (default) always attempt to invoke the task regardless of the number of previous executions
> - `once` only invoke this task once regardless of the number of references
> - `when_changed` only invokes the task once for each unique set of variables passed into the task

---

> ❓ If 3 tasks running in parallel call the same task, is there a way of queueing the "same task" so that it only run consecutively? This is needed in cases like an `apt-get` where only one instance can run at a time.

This would be solved with the v4 `serialGroups` feature mentioned above. Task uniqueness is determined by it's name _and_ vars. `serialGroups` only cares about the task by name when putting it into a "serial group". Thus, multiple instances of the same task, but with different vars, when they have the `serialGroups` field set will be serialized (queued).

**Reply by ProfessorManhattan** [reply]
1. I think that would be able to solve basically all of my questions in combination with feeding the variables with idea #2
2. I think the JSON approach is good. On the fork I linked to, this is how we are handling dynamically populated questions (since we integrated a question/answer library -- example: https://github.com/megabyte-labs/Bodega/blob/master/.config/taskfiles/ansible/Taskfile-playbook.yml#L111).
3. I don't know any Go so I don't understand --- please consider supporting `deps` still, it might be worthwhile to continue supporting `deps` and make that run first. Then, `needs` could run after the `env` and `vars` populate. Take this as an example: https://github.com/megabyte-labs/Bodega/blob/master/.config/taskfiles/docker/Taskfile-test.yml#L52 --- yq needs to be installed before the `status` runs but needs, being that it's going to be all dynamic and whatnot could be a second, more-flexible version of `deps` --- I think this introduces another problem that might be worthwhile to solve - it would be useful if we could somehow control the order in which `deps`, `vars`, `env` all run. At the very least, we should update the documentation to note the order in which the different steps all run.
4. Thanks for taking the time to respond to my questions

I have one final question. I don't know much Go but want to learn. For this project though, I hired a contractor to add customizations because I assumed development on this was slow but now that you're mentioning all these cool new features, I don't want to miss out. What's the best way of incorporating new features like:

1. A `task --menu` command that converts `task --list` into a more usable, filterable, prettier list by mimicing the functionality in Glow (https://github.com/charmbracelet/glow)
2. Rendering `task --summary` with Glamour (https://github.com/charmbracelet/glamour)
3. Adding a question/answer system (Example implementation: https://github.com/megabyte-labs/Bodega/blob/master/.config/taskfiles/ansible/Taskfile-playbook.yml#L111) by utilizing this library: https://github.com/AlecAivazis/survey -- I chose to use this library because it was one of the most popular and figured it would be straightforward to convert between YAML to Go syntax
4. Adding `run: once_system` which stores its cache in a global directory --- there should also probably be a `run: once` that stores its cache in the `.task` folder called `run: once_project` or something similar. The `run: once_system` might be used to run a task that adds the MAC address to a firewall API. The `run: once_project` might be used to run a full history Gitleaks scan.

There's a handful of other ones that we've already developed. I can open PRs but I'd like to know which ones you would be willing to include in advance since there's some work involved from my team to create the PR. I would understand why you would not want to include Glamour for styled markdown-compatible `task --summary`. Task is pretty minimal and incorporating libraries like this could add a few megabytes to the build size. Any ideas on how we can make Task pluggable? IMO, Task is awesome and is only going to get more and more popular. It would be great if all new features to Task dogfooded a well-thought, simple plugin system.

**Reply by ProfessorManhattan** [reply]
I would much rather contribute to this project over my custom fork but I'm not sure if these TUI usability features align with the owner's vision for task. Is it supposed to be as minimal as possible for mostly headless environments or is it supposed to be an easy-to-use, TUI experience (that's a few megs larger)?

Another thing we did that you might find interesting is we incorporated a websocket controller so that Task can be controlled from a local web app. If we ran Task as a daemon like this, we could potentially trigger tasks in a loop through a websocket that would not spin up another process. It takes away from the simplicity but it might be useful. Hopefully some of my teams POC work can make it into this project!

All of the features except number 1 are already implemented on the fork.

CC: @andreynering @ghostsquad

**Reply by ghostsquad** [reply]
> please consider supporting deps still

I didn't say I was removing `deps`, I said I was renaming `deps` to `needs`. I have found that using more verbs can really help when you are reading/writing files like this. It's why you see a lot of verbs in github actions workflow syntax.

---

> What's the best way of incorporating new features like

These are good questions. I'm still very much in the infancy of writing task v4, and there's a lot to consider. **Simplicity** is at the forefront of why I even started on a V4. I want the task API to feel fresh, and easy to contribute to. I want there to be a set of "rules", like in a video/board game, so that even if something isn't explicitly mentioned in the docs, you may still be able to "intuition" your way to an answer. So, I may not be doing _any_ advanced TUI stuff out of the gate, but I will be keeping it in mind, and I want to design the API to make this easy to implement and test. For now, I would say, just stay tuned, and if you have any concrete ideas, file an issue, and I can mark it as a `v4` feature request, and that's how I can track this stuff.

---

> Adding a question/answer system

I do like the idea of "wizards". I find that it can be great for new users and "discoverability".

---

> Adding run: once_system which stores its cache in a global directory

This is starting to get into the territory of something like `pulumi` or `terraform`. I will think more about this, but I think this is may be out of scope for `task`.

---

> Any ideas on how we can make Task pluggable? IMO, Task is awesome and is only going to get more and more popular. It would be great if all new features to Task dogfooded a well-thought, simple plugin system.

This is already possible actually. Task supports running other executables, and in `v4` I'm looking for more places to make that possible. Consider this:

```txt
my-project
├── Taskfile.yml
└── cmd
    └── thingdoer
        └── main.go
```

> `Taskfile.yml`

```yaml
tasks:
  thing:
    cmds:
    - go run cmd/thingdoer/main.go
```

You can literally drop right into `go` by simply using `go run`. Stdin, Stdout, Stderr (simple plugin?) are all available to you.

There's probably more to do in regards to variable passing though.

---

> I would much rather contribute to this project over my custom fork but I'm not sure if these TUI usability features align with the owner's vision for task. Is it supposed to be as minimal as possible for mostly headless environments or is it supposed to be an easy-to-use, TUI experience (that's a few megs larger)?

There are a lot of tools I use that mix command-line (headless) usage with "pretty" TUI functionality. Even `docker` does, and is doing this more and more with recent versions.

Regarding implementation, there are several ways to make this a reality, and "do the right thing" automatically. A simple example is by just detecting if the tool is running in a `tty` (https://github.com/mattn/go-isatty). Leveraging a user's local config files (`~/.config/task.yml`) to enable/disable user-experience features, without affecting the determinism of the tasks themselves. Finally, allowing command-line options, such as `--interactive=false` or like with `git` using `git --no-pager`.

I want task to be a pleasant user experience when it's in your face, and I want it to "just work" when running in a CI environment or when it's part of higher level automation. So, those are my design goals for V4.

Regarding size, K9s is 13-14mb CLI. But it's [absolutely gorgeous](https://github.com/derailed/k9s#screenshots)

And as you mentioned, all the charmbracelet stuff is very very nice, e.g [lipgloss](https://github.com/charmbracelet/lipgloss).

and more (https://github.com/avelino/awesome-go#advanced-console-uis)

Personally, I don't mind having a larger binary, but also, with a proper API design, it might be possible to generate 2 binaries, one that is very light, and doesn't support the advanced TUI features, which would mainly be used/installed in CI systems, and a larger binary for use by developers.

---

> Another thing we did that you might find interesting is we incorporated a websocket controller so that Task can be controlled from a local web app.

Explicit support for this would likely be outside the scope of the project, however, I am interested in `task` as an API. So if you wanted to write your own webserver that used the `task` API programmatically, that would certainly be pretty cool.

---

> All of the features except number 1 are already implemented on the fork.

I'll be taking a look at your code for inspiration. Thank you!

**Reply by ssbarnea** [reply]
I am really interested about matrix functionality. A very good example to learn from would be how github actions matrix jobs are defined, with multiple matrix variables and the include and exclude options for adding/substacting from matrix.

Another interesting place to learn from would be tox.

**Reply by ProfessorManhattan** [reply]
Hey @ghostsquad  --- I did some work and created a couple POCs... one of them has a TUI Task Dashboard that is based on Charm Bracelet's Glow library. We're discussing if / how we can get it merged into Task here: https://github.com/go-task/task/discussions/1077

**Reply by ghostsquad** [reply]
@ProfessorManhattan Sorry, task has taken a back seat for me. I've been focusing on other things recently. Additionally, @andreynering has changed some settings, so I'm not really even a maintainer anymore. I've been punished for merging a PR that broke something because no regression tests existed.

**Reply by ProfessorManhattan** [reply]
Hey @ghostsquad .. ok why not do quarterly releases or something like that? My company does biweekly releases so all the developers are walking on ice.

**Reply by ghostsquad** [reply]
It's not a release cadence that really matters. Stuff happens, bugs show up in the code. There simply needs to be a process and understanding for how to deal with it when it does happen instead expecting perfection on every merge.

**Reply by ProfessorManhattan** [reply]
Yeah.. I'm totally with you.. I personally think the master branch should be looked at like a beta branch and the release channels like Homebrew should be powered by a secondary branch.. not sure how your CI works. Still forming an opinion about it though, personally.

**Reply by ghostsquad** [reply]
The way I treat continuous deployment/delivery is that `main`/`master` is "what you want to go to production", but it must first pass all manual/automated tests/gates before that happens. The best way in `git` to then track things is with _tags_. It's more or less "GithubFlow" with a mix of "OneFlow". The mix happens when a hotfix is needed in the live running production environment. Only when a hotfix is needed, would you branch off of the current tagged commit, commit the hotfix and retag the new commit, followed by merging that back into main.

Here's a brief article for further reading.
https://betterprogramming.pub/a-simple-git-repository-strategy-93a0c7450f23

Maybe I should write a blog post about this. This makes the most sense for _deployable_ applications, but generally speaking, it can work with tools & libraries too. Creating a tag indicates a release in some form. That release could be a _beta_, _prerelease_, or _release candidate_. Which could then be tagged again if no additional work is needed to indicate that it's been tested/vetted.

The automated CI/CD process could then be watching for _tags_, and acting on the tag based on semver semantics.

---

A rollback can be handled in various ways. For something like this repo (i.e. tool/library), either "rolling forward" via a new commit that reverts a change, or a commit that fixes the problem are all that really needs to be done.
